sys/kern/proc.c

```diff
@@ -259,46 +261,68 @@ static pgrp_t *pgrp_lookup(pgid_t pgid) {
   return NULL;
 }
 
-/* Make process leaves its process group. */
+static void pgrp_remove(pgrp_t *pgrp) {
+  assert(mtx_owned(all_proc_mtx));
+
+  session_drop(pgrp->pg_session);
+  TAILQ_REMOVE(PGRP_HASH_CHAIN(pgrp->pg_id), pgrp, pg_hash);
+  pool_free(P_PGRP, pgrp);
+}
+
+/* Make a process leave its process group.
+ * Called only on process exit, use _pgrp_enter for changing groups. */
 static void pgrp_leave(proc_t *p) {
-  /* We don't want for any process to see that our p_pgrp is NULL. */
   assert(mtx_owned(all_proc_mtx));
 
   pgrp_t *pgrp = p->p_pgrp;
 
-  TAILQ_REMOVE(&p->p_pgrp->pg_members, p, p_pglist);
-  p->p_pgrp = NULL;
-
-  if (TAILQ_EMPTY(&pgrp->pg_members)) {
-    session_drop(pgrp->pg_session);
-    TAILQ_REMOVE(PGRP_HASH_CHAIN(pgrp->pg_id), pgrp, pg_hash);
-    pool_free(P_PGRP, pgrp);
+  WITH_MTX_LOCK (&pgrp->pg_lock) {
+    WITH_PROC_LOCK(p) {
+      TAILQ_REMOVE(&p->p_pgrp->pg_members, p, p_pglist);
+      p->p_pgrp = NULL;
+    }
   }
+
+  if (TAILQ_EMPTY(&pgrp->pg_members))
+    pgrp_remove(pgrp);
 }
 
 static int _pgrp_enter(proc_t *p, pgrp_t *target) {
+  pgrp_t *old_pgrp = p->p_pgrp;
+  assert(old_pgrp);
+
+  if (old_pgrp == target)
+    return 0;
+
   pgrp_increase_jobc(p, target);
```

Comment:
I got confused looking at those two lines. Without diving into `pgpr_*_jobc` implementation this looks almost like no-op. Did we choose good names for these procedures?

Reply:
Maybe `pgrp_jobc_enter` and `pgrp_jobc_leave` are better names? Honestly, I don't think this is a big deal. One can easily find the implementation of these procedures and read the comment that explains their purpose.

Reply:
We could also add `old_pgrp` as a parameter to `pgrp_decrease_jobc`. It would look less like a no-op then.

Reply:
I'll change it as suggested.